File: mm2.py
# -*- coding: utf8 -*-
# скрипт для ансамблирования глобальной матрицы жесткости и решения СЛАУ

import sys

# настройки
n = 41					# количество узлов
form = 1 								# 3 — кубическая функция формы, 1 — линейная функция формы

if len(sys.argv) == 3:
	n = int(sys.argv[1])
	form = int(sys.argv[2])
	if form != 1 and form != 3:
		exit('wrong form choice')
else:
	print "give first arg n, second — form (1 or 3)"
	exit()

segmentSize = 11.0		# длина рассматриваемого интервала

l = segmentSize / (n - 1)		# длина конечного элемента

print "l = ", l, "n = ", n, "form: ", form
# коэффициенты уравнения конечного элемента, когда известны первые производные на концах элемента


if form == 1:
	a11 = (5*l*l + 1) / l
	a12 = (5*l*l - 2) / (2*l)
	a21 = (5*l*l - 2) / (2*l)
	a22 = (5*l*l + 1) / l
	b1 = 2*l
	b2 = 2*l

if form == 3:
	v1 = (15*l**6 + 135*l**4 + 192*l**2 + 28) / (l*(15*l**4 + 52*l**2 + 28))		# переменные для удобства
	v2 = (15*l**6 - 30*l**4 + 72*l**2 - 112) / (4*l*(15*l**4 + 52*l**2 + 28))
	v3 = l*(l**2 + 4) / (3*l**2 + 2)
	a11 = v1
	a12 = v2
	a21 = v2
	a22 = v1
	b1 = v3
	b2 = v3

deriv1 = 123			# в решаемом варианте (23) это значение нам не известно
deriv2 = 1
u0 = 10
u_end = 123				# в решаемом варианте (23) это значение нам не известно

def printMatrix(a, b):
	idx = 0
	for row in a:
		print row, b[idx]
		idx += 1
	print " "

# Matrix = [[0 for x in xrange(5)] for x in xrange(5)] 

T = range(n)
F = range(n)
for i in range(n):
	T[i] = range(n)
	F[i] = 0

for i in range(n):
	for j in range(n):
		T[i][j] = 0

for i in range(n - 1):
	T[i][i] += a11
	T[i][i+1] += a12
	T[i+1][i] += a21
	T[i+1][i+1] += a22
	F[i] += b1
	F[i+1] = b2
F[0] += -deriv1
F[n-1] += deriv2 

# теперь, если решить СЛАУ TU = F относительно U, в U будут искомые значения функции в соответствующих точках при условии, что нам известны первые производные на концах интервала.
# так как первая производная (deriv1) на начале интервала нам не известна, но известно само значение (U[0]), можно исключить первый столбец матрицы T домножив его на U[0]
# и перенеся в правую часть (F); теперь можно исключить и первую строчку с неизвестной первой производной

#printMatrix(T, F)

# учет u0 в F
for i in range(1, n):
	F[i] -= T[i][0]*u0 
for i in range (n):		# удаление первого столбца
	T[i].pop(0)	
T.pop(0)				# удаление первой строчки
F.pop(0)

printMatrix(T, F)

# добавление строки и столбца
for i in range (n - 1):						# добавление столбца
	T[i].append(0)
T.append([0 for x in range(n)])				# добавление строчки
F.append(0)									# добвление единичной строчки к правой части

T[n - 2][n - 1] = -1
F[n - 2] = 0
T[n - 1][n - 2] = -1
T[n - 1][n - 1] = 1
F[n - 1] = 0

printMatrix(T, F)

from numpy import linalg
U = linalg.solve(T, F)
print U

# тест солвера слау
#total_error = 0
#for i in range(n-1):
#	sum = 0
#	for j in range(n-1):
#		sum += T[i][j] * U[j]
#	local_error = abs(F[i] - sum)
#	total_error += local_error
#print "total_error: ", total_error

# графика
from Tkinter import *
from math import sqrt, exp
root = Tk()
root.title('Model')

scale_k = 50
height = 600
width = 800
minus_height = 0

c = Canvas(root, height=height+minus_height, width=width)
c.pack()

def drawPoint(x, y, w,color):
	ex = w / 2				# эксцентриситет
	x = x * scale_k			# масштабирование
	y = y * scale_k
	x_p = 2 + x
	y_p = height - 2 - y	
	c.create_oval(x_p - ex, y_p - ex, x_p + w - ex, y_p + w - ex, outline=color, fill=color)

#def f(x):
#	return 2*(10**(-20))*exp(4*x) + 9.73*exp(-4*x) + 0.27

#def f(x):		# orig func
#	return (1/(15*(1 + exp(22*sqrt(15)))) ) * exp(-sqrt(15) * x) * (4*exp(sqrt(15)*x) + 146*exp(2*sqrt(15)*x) + 4*exp(sqrt(15)*(x+22)) + sqrt(15)*exp(sqrt(15)*(2*x+11)) - sqrt(15)*exp(11*sqrt(15)) + 146*exp(22*sqrt(15)))

def f(x):		# решение д.у. с граничным условием 3 рода на правой границе
	return (2*exp(-sqrt(15)*x) * (2*exp(sqrt(15)*x) + 2*sqrt(15)*exp(sqrt(15)*x) + 73*exp(2*sqrt(15)*x) + 73*sqrt(15)*exp(2*sqrt(15)*x) - 2*exp(sqrt(15)*x+22*sqrt(15)) + 2*sqrt(15)*exp(sqrt(15)*x+22*sqrt(15)) + 2*exp(2*sqrt(15)*x+11*sqrt(15)) - 2*exp(11*sqrt(15)) - 73*exp(22*sqrt(15)) + 73*sqrt(15)*exp(22*sqrt(15)))) / (15*(1 + sqrt(15) - exp(22*sqrt(15)) + sqrt(15)*exp(22*sqrt(15))))

y = 0.0
while(y <= segmentSize):
	drawPoint(0, y, 1,'black')
	y += 0.01

x = 0.0
while(x <= segmentSize):
	drawPoint(x, 0, 1,'black')
	drawPoint(x, f(x), 1,'red')
	x += 0.01

# отрисовка полученного решения и вывод численного решения
maxErr = 0.0
idxWithMaxErr = -1
drawPoint(0, u0, 7,'green')
for i in range(1, n):
	x = i*l
	currErr = abs(f(x) - U[i - 1])
	if currErr > maxErr:
		maxErr = currErr
		idxWithMaxErr = i
	print "x: %f\tf: %f.3\tu: %f.3\t\tdiff: %f" % (x, f(x), U[i - 1], currErr)
	drawPoint(x, U[i - 1], 7,'green')
print "maxErr: ", maxErr

mainloop()

File: mm2.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

void printm(double **m, double *f,int n)
{
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
		{
			printf("%f\t", m[i][j]);
			if (j == n - 1)
			{
				printf("\t%f\n", f[i]);
			}
		}
	}
	printf("\n");
}

double *gauss(double **m, double *f, int n)
{
	for (int globRowIdx = 0; globRowIdx < n; globRowIdx++)
	{
		// ведущий элемент для главной обрабатываемой строки должен быть равен 1
		double leadDevisor = m[globRowIdx][globRowIdx];
		for (int j = globRowIdx; j < n; j++)
		{
			m[globRowIdx][j] = m[globRowIdx][j] / leadDevisor;	
		}
		f[globRowIdx] = f[globRowIdx] / leadDevisor;

		for (int locRowIdx = globRowIdx + 1; locRowIdx < n; locRowIdx++)
		{
			double leadMultiplier = m[locRowIdx][globRowIdx];
			for (int j = globRowIdx; j < n; j++)
			{
				m[locRowIdx][j] -= leadMultiplier * m[globRowIdx][j];
			}
			f[locRowIdx] -= leadMultiplier * f[globRowIdx]; 
		}

		if (globRowIdx == 1)
		{
			//break;
		}
	}

	printm(m, f, n);

	double *x = calloc(n, sizeof(double));

	x[n - 1] = f[n - 1];

	for (int xIdx = n - 2; xIdx >= 0; xIdx--)
	{
		double sum = 0;
		for (int i = xIdx + 0; i < n; xIdx++)
		{
			//sum += x[i] * m[xIdx][i];
		}
		x[xIdx] = f[xIdx] - sum;
	}



	for (int i = 0; i < n; i++)
	{
		printf("%f ", x[i]);
	}
	printf("\n");

	return x;

	/*for globalProcessingRowIdx in range(m):
	leadDevisor = T[globalProcessingRowIdx][globalProcessingRowIdx]
	for processingColumnIdx in range(globalProcessingRowIdx, n):
		T[globalProcessingRowIdx][processingColumnIdx] = T[globalProcessingRowIdx][processingColumnIdx] / leadDevisor				# bring leading element to 1

	for localProcessingRowIdx in range(globalProcessingRowIdx + 1, m):
		leadMultiplier = T[localProcessingRowIdx][globalProcessingRowIdx]
		for j in range(globalProcessingRowIdx, n):
			T[localProcessingRowIdx][j] -= leadMultiplier * T[globalProcessingRowIdx][j]

	if globalProcessingRowIdx == 1:
		break*/
}

int main(int argc, char **argv)
{
	int n = 4;								// количество узлов

	double segmentSize = 11.0;				// длина рассматриваемого интервала

	double l = segmentSize / (n - 1);		// длина конечного элемента

	printf("l = %f\n", l);

											// коэффициенты уравнения конечного элемента, когда известны первые производные на концах элемента


	double a11;
	double a12;
	double a21;
	double a22;
	double b1;
	double b2;


	int form = 1; 							// 3 — кубическая функция формы, 1 — линейная функция формы
	if (form == 1)
	{
		a11 = (5*l*l + 1) / l;
		a12 = (5*l*l - 2) / (2*l);
		a21 = (5*l*l - 2) / (2*l);
		a22 = (5*l*l + 1) / l;
		b1 = 2*l;
		b2 = 2*l;
	}

	if (form == 3)
	{
		double v1 = (15*pow(l, 6) + 135*pow(l, 4) + 192*pow(l, 2) + 28) / (l*(15*pow(l, 4) + 52*pow(l, 2) + 28));		// переменные для удобства
		double v2 = (15*pow(l, 6) - 30*pow(l, 4) + 72*pow(l, 2) - 112) / (4*l*(15*pow(l, 4) + 52*pow(l, 2) + 28));
		double v3 = l*(pow(l, 2) + 4) / (3*pow(l, 2) + 2);
		double a11 = v1;
		double a12 = v2;
		double a21 = v2;
		double a22 = v1;
		double b1 = v3;
		double b2 = v3;
	}

	double deriv1 = 123;			// в решаемом варианте (23) это значение нам не известно
	double deriv2 = 1;
	double u0 = 10;
	double u_end = 123;				// в решаемом варианте (23) это значение нам не известно

	printf("hello %f\n", a11);

	// создание матрицы
	double **matrix = calloc(n, sizeof(double **));
	for (int idx = 0; idx < n; idx++)
	{
		matrix[idx] = calloc(n, sizeof(double));
	}

	double *f = calloc(n, sizeof(double));

	// инициализация матрицы
	for (int i = 0; i < n - 1; i++)
	{
		matrix[i][i] += a11;
		matrix[i][i+1] += a12;
		matrix[i+1][i] += a21;
		matrix[i+1][i+1] += a22;
		f[i] += b1;
		f[i+1] = b2;
	}	
	f[0] += -deriv1;
	f[n-1] += deriv2 ;

	matrix[0][0] = 22.81850649;
	matrix[0][1] = -13.89776786;
	matrix[0][2] = -8.104748376;
	matrix[0][3] = 2.277759740;
	f[0] = 0.825;

	matrix[1][0] = -13.89776786;
	matrix[1][1] = 22.81850649;
	matrix[1][2] = 2.277759740;
	matrix[1][3] = -8.104748376;
	f[1] = 0.825;

	matrix[2][0] = -8.104748376;
	matrix[2][1] = 2.277759740;
	matrix[2][2] = 7.355844156;
	matrix[2][3] = -0.4976055194;
	f[2] = 0.275;

	matrix[3][0] = 2.277759740;
	matrix[3][1] = -8.104748376;
	matrix[3][2] = -0.4976055194;
	matrix[3][3] = 7.355844156;
	f[3] = 0.275;


	printm(matrix, f, n);

	// удаляем первый столбец и строку матрицы
	/*
	for (int i = 0; i < n - 1; i++)
	{
		memmove(matrix[i], matrix[i + 1] + 1, (n - 1)*sizeof(double));
	}
	memmove(f, f + 1, (n - 1)*sizeof(double));
	*/

	//printm(matrix, f, n);

	gauss(matrix, f, n);

	printm(matrix, f, n);

	for (int i = 0; i < n; i++)
	{
		free(matrix[i]);		
	}
	free(f);
}